iroh/src/magicsock/remote_map/remote_state/path_state.rs

```diff
@@ -123,4 +165,338 @@ pub(super) struct PathState {
     /// We keep track of only the latest [`Instant`] for each [`Source`], keeping the size
     /// of the map of sources down to one entry per type of source.
     pub(super) sources: HashMap<Source, Instant>,
+    /// The last time this path was proven usable.
+    ///
+    /// If this is `false` and the `abandoned` field is `Some`, than we attempted
+    /// to open this path, but it did not work.
+    ///
+    /// If this is `false` and the `abandoned` field is `None`, than we do not know
+    ///  yet if this path is usable.
+    pub(super) usable: bool,
+    /// The last time a path with this addr was abandoned.
+    ///
+    /// If this is `Some` and usable is `None`, than we attempted to use this path and it
+    /// did not work.
+    pub(super) abandoned: Option<Instant>,
+}
+
+/// Prunes the IP paths in the paths HashMap.
+///
+/// Only prunes if the number of IP paths is above [`MAX_IP_PATHS`].
+///
+/// Keeps paths that are active or have never been holepunched.
+///
+/// Always prunes paths that have unsuccessfully holepunched.
+///
+/// Keeps [`MAX_INACTIVE_IP_PATHS`] of the most recently closed paths
+/// that are not currently being used but have successfully been
+/// holepunched previously.
+///
+/// This all ensures that:
+///
+/// - We do not have unbounded growth of paths.
+/// - If we have many paths for this remote, we prune the paths that cannot hole punch.
+/// - We do not prune holepunched paths that are currently not in use too quickly. For example, if a large number of untested paths are added at once, we will not immediately prune all of the unused, but valid, paths at once.
+fn prune_ip_paths(paths: &mut FxHashMap<transports::Addr, PathState>) {
+    // if the total number of paths is less than the max, bail early
+    if paths.len() < MAX_IP_PATHS {
+        return;
+    }
+
+    let ip_paths: Vec<_> = paths.iter().filter(|(addr, _)| addr.is_ip()).collect();
+
+    // if the total number of ip paths is less than the max, bail early
```

Comment:
We are trying to get away from making IP paths special, but rather referring to direct paths, so maybe use that naming?

Reply:
Let's keep it like this for now. IP paths are still special because of holepunching.